React/testing.md

```diff
@@ -0,0 +1,234 @@
+ › Press Enter to trigger a test run.
+ ```
+
+The test-runner will automatically run tests which have changed since the last save.  As you develop a React application, you can leave the tests running in a terminal tab and the tests will re-execute when you save files.
+
+  **Press `a` now** to manually re-run all the existing tests.
+
+ You should get the following:
+ ```bash
+ ... big error message
+
+Test Suites: 1 failed, 1 total
+Tests:       1 failed, 1 total
+Snapshots:   0 total
+Time:        3.34s
+Ran all test suites.
+
+Watch Usage: Press w to show more.
+```
+
+**OMG!!!** We're already failing a test!   That test file is `src/App.test.js`.
+
+Open that file:
+
+```javascript
+// src/App.test.js
+
+import React from 'react';
+import { render } from '@testing-library/react';
+import App from './App';
+
+test('renders learn react link', () => {
+  const { getByText } = render(<App />);
+  const linkElement = getByText(/learn react/i);
+  expect(linkElement).toBeInTheDocument();
+});
+```
+
+The test manually renders the `App` component and then tries to get some text with the words "learn react".
+
+Hmmm... we deleted that text...  
+
+**Question**  What do you notice?  What do you think is happening here?
+
+<details>
+  <summary>Walkthrough</summary>
+1.  The default test in `App.test.js` first renders an instance of the `App` component in a simulated DOM.  
+2.  Once the component is rendered it returns a `RenderResult` object which has a bunch of properties.
+3.  We pull the `getByText` function from the `RenderResult` by using [destructuring](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Destructuring_assignment).
+4.  Then we try to find an element inside the `App` component with the regular expression `/learn react/i`.
+5.  Lastly we have an expectation that element exists... but it doesn't
+</details>
+
+Modify the test so that it searches for text which we **have** added to the App component.
+
+## Snapshot Testing
+
+### What is it?
+
+With React we often want to ensure that a component is rendering properly.  For our `NewStudentForm` component we could write tests ensure that all the fields are rendering through tests that:
+
+1.  Ensure that the component renders 2 input fields.
+2.  Each field is labeled properly.
+3.  Ensure there is a submit button
+4.  and more and more and more...
+
+However these tests are pretty fragile, easy to break when you make minor changes to the form and tedious to update for UI refactoring.
+
+Instead we can:
+
+1.  Convert the rendered html of the component into a file with the file holding the most significant features of the component.
+2.  Compare future runs of the test to that snapshot and alert the user to any changes.
+
+When a component changes, we can be alerted and make corrections, or update the snapshot for future test runs.
+
+**Question:** Can you do TDD with a Snapshot test?
+
+**Question:** When would a snapshot test be useful?
```

Comment:
I think these are great questions! I think with the online format it may be helpful to have hidden answers to these as well.